### PR TITLE
Update API to return 'source_id' for users

### PR DESCRIPTION
**Pull request by tobiasbp**
Using the API, a user's _source_id_ can be set in the _CreateUserOption_ model, but the field is not returned in the _User_ model. 

This PR updates the _User_ model to include the field _source_id_ (The ID of the Authentication Source). 

---
The name _source_id_ is a bit confusing since it does not hint at any relation with the _login_name_ field. Perhaps _login_source_id_ or _login_id_ would be a better name? If a different name than _source_id_ was to be used, the name should also be updated in the _CreateUserOption_ and _EditUserOptions_ models.

### CODE REVIEW

**Comment by lunny** [comment]
You need to generate swagger docs.

File: services/convert/user.go
Comment (wxiaoguang) [inline services/convert/user.go:78]: I can see other fields are all 1-vs-1 mapping to user's field (eg: IsAdmin, LoginName, LastLogin, Language, IsActive, etc)

But why this field's name (SourceID) is different from the user's field (LoginSource)?

For consistency, I would prefer to use the same name.

Comment (tobiasbp) [reply services/convert/user.go:78]: I agree with you, but as I have described in the PR, the use of the name _sourceID_ for the field is already being used in other places like
[Swagger CreateUserOption](https://github.com/go-gitea/gitea/blob/2033eb7c1138a06ea052e6c6e6e5799187519e16/templates/swagger/v1_json.tmpl#L19063)
[CreateUserOption struct](https://github.com/go-gitea/gitea/blob/2033eb7c1138a06ea052e6c6e6e5799187519e16/modules/structs/admin_user.go#L11)

I could change to *result.LoginSource = user.LoginSource* and then map that to the confusing, but used for consistency, _source_id_ in the swagger file? I guess that would be better.

Let me know what you think. Thanks.

Comment (wxiaoguang) [reply services/convert/user.go:78]: Hmm ... good point, that's a legacy problem .... so it's better to keep using "SourceID"

**Review by wxiaoguang — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by techknowlogick** [comment]
I agree re `The name source_id is a bit confusing since it does not hint at any relation with the login_name field`, but I think, at least for now, keeping it the same is fine, especially since you have it documented in the swagger docs to denote the relationship.

**Comment by delvh** [comment]
Erm… Wait a sec.
Aren't we modifying the API right now?
In other words, we won't be able to undo this change.

**Comment by lunny** [comment]
> Erm… Wait a sec. Aren't we modifying the API right now? In other words, we won't be able to undo this change.

For API, if we really want to change, we can deprecate one and ref to a new one.

**Comment by tobiasbp** [comment]
I have [a PR for updating the SDK as well](https://gitea.com/gitea/go-sdk/pulls/654). Could you take a look @lunny ?

**Comment by lunny** [comment]
@delvh Do you have any concerns about merging this PR?

**Comment by lunny** [comment]
> I have [a PR for updating the SDK as well](https://gitea.com/gitea/go-sdk/pulls/654). Could you take a look @lunny ?

I will do that after this PR merged.

**Review by techknowlogick — CHANGES_REQUESTED** [review]
blocking on behalf of @delvh

**Comment by delvh** [comment]
Just to be clear, I don't care which option you decide on.
I don't know enough about this PR/this specific part of the codebase to evaluate that.
I'm simply against introducing something to the API when there is already discussion of doing it another way.

**Comment by techknowlogick** [comment]
@delvh to clarify, you are referencing the discussion above about possibly changing this value in the future to be more clear in what it means? And that instead of making the breaking change in the future, it should just be made in this PR before merging?
(If yes, I am for that as well. I was ok with this going in above, as I felt this was an ok middlestep, but I understand not wanting a breaking change, and taking that extra effort now).

**Comment by tobiasbp** [comment]
> @delvh to clarify, you are referencing the discussion above about possibly changing this value in the future to be more clear in what it means? And that instead of making the breaking change in the future, it should just be made in this PR before merging? (If yes, I am for that as well. I was ok with this going in above, as I felt this was an ok middlestep, but I understand not wanting a breaking change, and taking that extra effort now).

How do you all feel about returning *login_source_id* (instead of *source_id*), and then also add *login_source_id* as an alias of *source_id* in the _CreateUserOption_ and _EditUserOptions_ models?

**Comment by delvh** [comment]
>  to clarify, you are referencing the discussion above about possibly changing this value in the future to be more clear in what it means? And that instead of making the breaking change in the future, it should just be made in this PR before merging?

Yes

**Comment by tobiasbp** [comment]
> How do you all feel about returning _login_source_id_ (instead of _source_id_), and then also add _login_source_id_ as an alias of _source_id_ in the _CreateUserOption_ and _EditUserOptions_ models?

What needs to happen to move forward with this PR? What do you think about what I proposed earlier?

**Comment by tobiasbp** [comment]
Is there no way to get this into 1.22.0 @lunny? I don't know what I need to change to make @techknowlogick and @delvh approve of it?

**Comment by lunny** [comment]
This looks not too big and is not a feature and got 2 approvals before the v1.22 branch was created. I think it's possible to backport to v1.22 if @techknowlogick and @delvh have no objections.

**Comment by delvh** [comment]
So, did you come to a conclusion which name to use?

**Comment by tobiasbp** [comment]
> So, did you come to a conclusion which name to use?

I proposed this earlier: https://github.com/go-gitea/gitea/pull/29718#issuecomment-2003145280

**Comment by tobiasbp** [comment]
So, what do you think of my suggestion @delvh @techknowlogick ?
https://github.com/go-gitea/gitea/pull/29718#issuecomment-2003145280

**Comment by lunny** [comment]
Suddenly, I have another idea. Since we have the opportunity to refactor the user system. Maybe we should remove login_source_id in the user table but put it into external_account because one user could bind multiple auth systems. So if we merge this PR, we need to deprecate it later.

**Comment by tobiasbp** [comment]
> Suddenly, I have another idea. Since we have the opportunity to refactor the user system. Maybe we should remove login_source_id in the user table but put it into external_account because one user could bind multiple auth systems. So if we merge this PR, we need to deprecate it later.

I'm not sure we are talking about the same thing here. A user can only have a single _authentication source_ .

image

**Comment by lunny** [comment]
This is the current design but I think the user can also login with some OAuth2.

**Comment by lunny** [comment]
> > This is the current design but I think the user can also login with some OAuth2.
> 
> I have this set up with two _openID_ _authentication sources_ for two different types of users. Since I can not get the _auth source ID_ from the API, I can not (easily) tell the users apart.
> 
> Is this system going to change in the next version of Gitea?

I think it will not be in next version of Gitea but we know it has the problem and would like to do a new design like that.

Maybe we can have a new API to get a user's all auth sources so it will look compatible with backend.
Sorry for the confusions.

**Comment by tobiasbp** [comment]
> I think it will not be in next version of Gitea but we know it has the problem and would like to do a new design like that.

Hmmm.... I'm now really confused @lunny . I think you are saying one of these things:

1. The next release of Gitea (1.22.0) will have the use of _authentication sources_ redesigned in a way that will allow users to have more than one _authentication source_. This feature changes the API in way that will make this PR obsolete. We should reject this PR.

2. Let's update the API for Gitea 1.22.0 in this request to add an endpoint that will return a users _authentication source_ in a format that will be compatible with the plans to introduce the feature of users having several _authentication sources_  in a version of Gitea beyond 1.22.0.

Thanks

**Comment by tobiasbp** [comment]
Sorry to be pushy, but what are we going to do about this _PR_? I think it's quite the oversight, that you can currently use the *API* to set the *source_id* for a user, but not retrieve it. @lunny @delvh @techknowlogick

**Comment by techknowlogick** [comment]
Yes, I agree. I think if we have a set, we need to have a get. And if/when a refactor happens then it'll also affect the set, so it'd already be breaking, so adding the get wouldn't affect things.
And thinking back about it, while a rename on the key would be nice, since it aligns with an already existing key for the set, then it would be probably best to keep the key name the same.

**Comment by tobiasbp** [comment]
> Yes, I agree. I think if we have a set, we need to have a get. And if/when a refactor happens then it'll also affect the set, so it'd already be breaking, so adding the get wouldn't affect things. And thinking back about it, while a rename on the key would be nice, since it aligns with an already existing key for the set, then it would be probably best to keep the key name the same.

Sounds good. Can you approve the PR please?

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**
